### ISSUE TITLE
I think the implementation of bimpm_mathcing is wrong

### ISSUE THREAD

**Issue by WANGZhaowei-Wesley** [allenai/allennlp#5260]
## Checklist

- [ ] I have verified that the issue exists against the `main` branch of AllenNLP.
- [x] I have read the relevant section in the [contribution guide](https://github.com/allenai/allennlp/blob/main/CONTRIBUTING.md#bug-fixes-and-new-features) on reporting bugs.
- [ ] I have checked the [issues list](https://github.com/allenai/allennlp/issues) for similar or identical bug reports.
- [ ] I have checked the [pull requests list](https://github.com/allenai/allennlp/pulls) for existing proposed fixes.
- [ ] I have checked the [CHANGELOG](https://github.com/allenai/allennlp/blob/main/CHANGELOG.md) and the [commit log](https://github.com/allenai/allennlp/commits/main) to find out if the bug was already fixed in the main branch.
- [ ] I have included in the "Description" section below a traceback from any exceptions related to this bug.
- [ ] I have included in the "Related issues or possible duplicates" section beloew all related issues and possible duplicate issues (If there are none, check this box anyway).
- [ ] I have included in the "Environment" section below the name of the operating system and Python version that I was using when I discovered this bug.
- [ ] I have included in the "Environment" section below the output of `pip freeze`.
- [ ] I have included in the "Steps to reproduce" section below a minimally reproducible example.

## Description
I read the code source code of Bi-MPM model(https://github.com/allenai/allennlp/blob/main/allennlp/modules/bimpm_matching.py) and I found the Attentive-Matching in the code is very different from what is described in the original paper. 
For example,  the 350, 351 lines, softmax is used to the hidden_state dimension(the last dimension). However, to be the same as what is in the paper, I think we just need to devide the weighted sum with the sum of weights.

<details>
<summary><b>Python traceback:</b></summary>
<p>

```
```

</p>
</details>

## Related issues or possible duplicates

- None

## Environment

OS:

Python version:

<details>
<summary><b>Output of <code>pip freeze</code>:</b></summary>
<p>

```
```

</p>
</details>

## Steps to reproduce

<details>
<summary><b>Example source:</b></summary>
<p>

```
```

</p>
</details>

**Comment by epwalsh** [comment]
I don't have much context here, but maybe @handsomezebra would like to weigh in.

**Comment by handsomezebra** [comment]
@WANGZhaowei-Wesley thanks for the comments on the implementation. Since I am not the author of the original paper so there might be some differences to the implementation in the paper. Please help to suggest the code, so we can compare, and also we can actually run the algorithm and evaluate the difference in the results.

**Comment by handsomezebra** [comment]
@WANGZhaowei-Wesley I looked at my code change history, my original code of the two lines looks like this:
```
            att_mean_h = div_safe(att_h.sum(dim=2), cosine_sim.sum(dim=2, keepdim=True))
            att_mean_p = div_safe(att_p.sum(dim=1), cosine_sim.sum(dim=1, keepdim=True).permute(0, 2, 1))
```

That probably is what you were talking about. For some reason I chose the current code, probably because of performance reason or readability reason which I don't remember exactly. But I think I did the training/testing using them and didn't find any problem after the change. Thanks!
